File: 179-largest-number/179-largest-number.py
class Solution:
    def largestNumber(self, nums: List[int]) -> str:
        def compare(num1, num2):
            concat1 = num1 + num2
            concat2 = num2 + num1
            if concat2 > concat1:
                return 1
            elif concat2 == concat1:
                return 0
            else:
                return -1
        
        for i in range(len(nums)):
            nums[i] = str(nums[i])
        nums = sorted(nums, key=functools.cmp_to_key(compare))
        if nums[0] == "0":
            return "0"
        else:
            ans = ""
            for num in nums:
                ans += num
            return ans

File: 179-largest-number/179-largest-number.java
class Solution {
    public String largestNumber(int[] nums) {
        Integer[] numsCopy = new Integer[nums.length];
        for (int i = 0; i < nums.length; i++) {
            numsCopy[i] = nums[i];
        }
        
        Arrays.sort(numsCopy, (a, b) -> {
            String x = String.valueOf(a);
            String y = String.valueOf(b);
            Long leftRight = Long.valueOf(x + y);
            Long rightLeft = Long.valueOf(y + x);
            if (leftRight == rightLeft) return 0;
            if (leftRight > rightLeft) return -1;
            return 1;
        });
        
        StringBuilder s = new StringBuilder();
        if (numsCopy[0] == 0) return "0";
        for (int num : numsCopy) {
            s.append(num);
        }
        return s.toString();
    }
}